designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,338 @@
+
+Each scenario has one or more relevant user experiences. The different debugging functionalities are discussed first. Then, the various user experiences are discussed, along with which scenarios they apply to.
+
+## What can be Debugged Locally
+
+### <a id="sam-template-resource-local"></a> SAM Template Resources (Local Invoke)
+
+SAM Template resources of type [`AWS::Serverless::Function`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html) represent Lambda functions. Lambda function code referenced by these resources can be locally Run or Debugged. The Toolkit uses SAM CLI to invoke the Lambda function, emulating how the function is run on AWS. A debugger can be attached to the invoked Lambda function code, and the event passed into the Lambda function can be customized.
+
+### <a id="sam-template-resource-api-gateway"></a> SAM Template Resources (API Gateway style Local Invoke)
+
+SAM Template resources that contain an event of type [Api](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-api.html), can be hosted in a web server for local development and iteration. A REST request causes the Lambda function to receive an API Gateway based event.
+
+At this time, an experience for API Gateway support is not available. Some portions of this document (Debug Configurations for example) have considerations made for supporting API Gateway functionality.
+
+### <a id="standalone-lambda"></a> Standalone Lambda Function Handlers
+
+Lambda function handler code can be locally Run or Debugged independent of any SAM Application. This functionality is powered by SAM in order to provide a host to the Lambda hander. The Toolkit produces a temporary SAM Application that references the handler code. This temporary SAM Application is handled as mentioned [earlier](#sam-template-resource-local). At the end of the debug session, the temporary SAM Application is removed.
+
+In this mode, any SAM Templates that reference a handler are ignored. This prevents confusion/errors introduced when trying to perform a reverse-lookup between code and SAM Template resources (examples include incorrectly determining a function's Lambda handler string, or situations where more than one resource references the same function).
+
+The Toolkit does not provide support for locally running or debugging standalone Lambda function handlers as API Gateway calls. The code should be referenced from a SAM Template in order to use the API Gateway style debugging features mentioned in the earlier section.
+
+## <a id="debug-config"></a> What can be configured for a Debug session?
+
+The following parameters influence a debug session. These are user-configured, and are referenced by the various [debugging experiences](#debugging-experiences).
```

Comment:
Additionally, it would be useful for the user to be able to load / utilize their [SAM CLI config - sam.conf (toml)](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-config.html).

This lets the user switch seamlessly between SAM CLI and the IDE without having to duplicate configurations (which can they go out of sync).

---

The experience could be that we auto-load this file if present (and also allow the user to configure the file location explicitly).

We can then pre-populate the settings from there and let the user configure overrides.
Alternatively, we could require use of that file as the source of truth to avoid confusion. This seems similar to how we handle AWS credentials.